src/View/ViewLayerData.php

```diff
@@ -0,0 +1,254 @@
+            throw new InvalidArgumentException('$data must not be null');
+        }
+        if ($data instanceof ViewLayerData) {
+            $data = $data->data;
+        } else {
+            $source = $source instanceof ModelData ? $source : null;
+            $data = CastingService::singleton()->cast($data, $source, $name);
+        }
+        $this->data = $data;
+    }
+
+    /**
+     * Needed so we can rewind in SSViewer_Scope::next() after getting itemIteratorTotal without throwing an exception.
+     */
+    public function getIteratorCount(): int
+    {
+        $count = $this->getRawDataValue('count');
+        if (is_numeric($count)) {
+            return $count;
+        }
+        if (is_countable($this->data)) {
+            return count($this->data);
+        }
+        if (ClassInfo::hasMethod($this->data, 'getIterator')) {
+            return iterator_count($this->data->getIterator());
+        }
+        return 0;
+    }
+
+    public function getIterator(): Traversable
+    {
+        if (!is_iterable($this->data) && !ClassInfo::hasMethod($this->data, 'getIterator')) {
+            $type = get_class($this->data);
+            throw new BadMethodCallException("$type is not iterable.");
+        }
+
+        $iterator = $this->data;
+        if (!is_iterable($iterator)) {
+            $iterator = $this->data->getIterator();
+        }
+        $source = $this->data instanceof ModelData ? $this->data : null;
+        foreach ($iterator as $item) {
+            yield $item === null ? null : ViewLayerData::create($item, $source);
+        }
+    }
+
+    /**
+     * Checks if a field is set, or if a getter or a method of that name exists.
+     * We need to check each of these, because we don't currently distinguish between a property, a getter, and a method
+     * which means if any of those exists we have to say the field is "set", otherwise template engines may skip fetching the data.
+     */
+    public function __isset(string $name): bool
+    {
+        // Note we explicitly DO NOT call count() or exists() on the data here because that would
+        // require fetching the data prematurely which could cause performance issues in extreme cases
+        return in_array($name, ViewLayerData::META_DATA_NAMES)
+            || isset($this->data->$name)
+            || ClassInfo::hasMethod($this->data, "get$name")
+            || ClassInfo::hasMethod($this->data, $name);
+    }
+
+    public function __get(string $name): ?ViewLayerData
+    {
+        $value = $this->getRawDataValue($name, type: ViewLayerData::TYPE_PROPERTY);
+        if ($value === null) {
+            return null;
+        }
+        $source = $this->data instanceof ModelData ? $this->data : null;
+        return ViewLayerData::create($value, $source, $name);
+    }
+
+    public function __call(string $name, array $arguments = []): ?ViewLayerData
+    {
+        $value = $this->getRawDataValue($name, $arguments, ViewLayerData::TYPE_METHOD);
```

Comment:
```suggestion
        $value = $this->getRawDataValue($name, $arguments);
```

Reply:
Discussion about this can be in https://github.com/silverstripe/silverstripe-framework/pull/11405#discussion_r1815944022

Reply:
Done